### PR TITLE
fix(barcode): use draw buffer to set canvas to meet align requirement 

### PR DESCRIPTION
**Pull request by XuNeo**
### Description of the feature or fix

Continue of #5012 #5022 
Only with draw buffer interface, the start address and stride alignment can meet requirement.

### Checkpoints
- [ ] Run `code-format.py` from the scripts folder. [astyle](http://astyle.sourceforge.net/install.html) needs to be installed.
- [ ] Update the [Documentation](https://github.com/lvgl/lvgl/tree/master/docs) if needed
- [ ] Add [Examples](https://github.com/lvgl/lvgl/tree/master/examples) if relevant. 
- [ ] Add [Tests](https://github.com/lvgl/lvgl/blob/master/tests/README.md) if applicable.
- [ ] If you added new options to `lv_conf_template.h` run [lv_conf_internal_gen.py](https://github.com/lvgl/lvgl/blob/release/v8.3/scripts/lv_conf_internal_gen.py) and update [Kconfig](https://github.com/lvgl/lvgl/blob/release/v8.3/Kconfig).
 
Be sure the following conventions are followed:
- [ ] Follow the [Styling guide](https://github.com/lvgl/lvgl/blob/master/docs/CODING_STYLE.md)
- [ ] Prefer `enum`s instead of macros. If inevitable to use `define`s export them with `LV_EXPORT_CONST_INT(defined_value)` right after the `define`.
- [ ] In function arguments prefer `type name[]` declaration for array parameters instead of `type * name`
- [ ] Use typed pointers instead of `void *` pointers
- [ ] Do not `malloc` into a static or global variables. Instead declare the variable in `lv_global_t` structure in [`lv_global.h`](https://github.com/lvgl/lvgl/blob/master/src/core/lv_global.h) and mark the variable with `(LV_GLOBAL_DEFAULT()->variable)` when it's used. See a detailed description [here](https://docs.lvgl.io/master/get-started/bindings/micropython.html#memory-management).
- [ ] Widget constructor must follow the `lv_<widget_name>_create(lv_obj_t * parent)` pattern.
- [ ] Widget members function must start with `lv_<module_name>` and should receive `lv_obj_t *` as first argument which is a pointer to widget object itself.  
- [ ] `struct`s should be used via an API and not modified directly via their elements.
- [ ] `struct` APIs should follow the widgets' conventions. That is to receive a pointer to the `struct` as the first argument, and the prefix of the `struct` name should be used as the prefix of the function name too (e.g.  `lv_disp_set_default(lv_disp_t * disp)`)
- [ ] Functions and `struct`s which are not part of the public API must begin with underscore in order to mark them as "private".
- [ ] Arguments must be named in H files too.
- [ ] To register and use callbacks one of the following needs to be followed (see a detailed description [here](https://docs.lvgl.io/master/get-started/bindings/micropython.html#callbacks)):
  - For both the registration function and the callback pass a pointer to a `struct` as the first argument. The `struct` must contain `void * user_data` field.
  - The last argument of the registration function must be `void * user_data` and the same `user_data` needs to be passed as the last argument of the callback.
  - Callback types not following these conventions should end with `xcb_t`.

### CODE REVIEW

File: src/widgets/canvas/lv_canvas.h [hunk 1]
```diff
@@ -63,7 +63,9 @@ lv_obj_t * lv_canvas_create(lv_obj_t * parent);
  * @param h height of the canvas
  * @param cf color format. `LV_COLOR_FORMAT...`
  */
-void lv_canvas_set_buffer(lv_obj_t * canvas, void * buf, int32_t w, int32_t h, lv_color_format_t cf);
+void lv_canvas_set_buffer(lv_obj_t * obj, void * buf, int32_t w, int32_t h, lv_color_format_t cf);
+
+void lv_canvas_set_draw_buf(lv_obj_t * obj, lv_draw_buf_t * draw_buf);
```
Comment (kisvegabor) [inline src/widgets/canvas/lv_canvas.h:76]: Please add comments here and for `lv_canvas_get_draw_buf`.

**Review by kisvegabor — COMMENTED** [review]
I like the approach, especially because the `lv_canvas_set_buffer` is available too.

For consistency `lv_canvas_set_draw_buffer` might be a better name, but it's doesn't matches `lv_draw_buf_t`.

What about a simple search replace of `draw_buf` to `draw_buffer`?

**Comment by kisvegabor** [comment]
`lv_image_create()` is already used by the `lv_image` widget, however `lv_image_buf_*` could work. 

It really opens to door for some simplification and removal of duplication. 

Let's try it, if it really looks good we are still on time to add this change. 

I merge it now, to no block the development of with nitpicking on naming.

**Comment by kisvegabor** [comment]
Just to confirm: I'll merge it by keeping the 3 commits. Ok?

**Comment by XuNeo** [comment]
@kisvegabor @FASTSHIFT Please help to review the merge the code.
Please do not squash the two commits if that's OK.

File: src/libs/barcode/lv_barcode.c [hunk 2]
```diff
@@ -195,43 +195,28 @@ static void lv_barcode_destructor(const lv_obj_class_t * class_p, lv_obj_t * obj
...
-        return;
-    }
-
-    LV_LOG_INFO("free canvas buffer: %p", lv_canvas_get_buf(obj));
+    lv_draw_buf_t * draw_buf = lv_canvas_get_draw_buf(obj);
+    if(draw_buf == NULL) return;
 
-    lv_draw_buf_free((void *)lv_canvas_get_buf(obj));
-    img->data = NULL;
+    lv_draw_buf_destroy(draw_buf);
```
Comment (kisvegabor) [inline src/libs/barcode/lv_barcode.c:201]: Shouldn't it be handled in `lv_canvas`?

Comment (XuNeo) [reply src/libs/barcode/lv_barcode.c:201]: The draw buffer is allocated in `bardcode`, thus should be destroyed in `barcode` too. Canvas could also use static draw buffer, thus cannot destroy externally provided buffer.

**Pull request merged**